rs_exp(): prec converted to int before entering for loop

ishanaj opened the pull request:
Fixes issue #10191 
The statement:
```
>>>rs_series(sin(a**QQ(3,7))*exp(a + b**QQ(6,7)), a,2)
```
 produces a `TypeError` which is because the `for` loop in the function `rs_expr()` has `prec` as its range, while `prec` sometimes stores a fraction value, which thus generates the error.
So on evaluating the fraction and converting it to `int` the error disappears.
Now, for the missing term as stated in the issue, the value of `prec` is incremented by one.

<!-- BEGIN RELEASE NOTES -->
* polys
  * `rs_series` no more produces `TypeError` for valid inputs.
<!-- END RELEASE NOTES -->

ishanaj committed:
rs_exp: prec converted to int before entering for loop

Yathartha22 commented on sympy/polys/tests/test_ring_series.py:
```diff
@@ -629,3 +629,6 @@ def test_rs_series():
                     x**2*a**4/2 + x*a**2
+    assert rs_series(sin(a**QQ(3,7))*exp(a + b**QQ(6,7)), a,2).as_expr() == a**QQ(10,7)*exp(b**QQ(6,7)) - \
+                    a**QQ(9,7)*exp(b**QQ(6,7))/6 + \
+                    a**QQ(3,7)*exp(b**QQ(6,7))                         #issue 10191
```
There should be a line at the end of file as per pep8.

ishanaj committed:
Improved quality

gxyd commented on sympy/polys/ring_series.py:
```diff
@@ -1140,6 +1140,8 @@ def rs_exp(p, x, prec):
     n = 1
     k = 1
     c = []
+    if not(type(prec) is int):
```
use isinstance instead.

ishanaj committed:
modified if statement to use isinstance

gxyd commented:
@shivamvats will you be able to review this? I don't have know much about this.

gxyd commented on sympy/polys/ring_series.py:
```diff
@@ -1140,6 +1140,8 @@ def rs_exp(p, x, prec):
     c = []
+    if not isinstance(prec, int):
+        prec = int(prec.p/prec.q) + 1
     for k in range(prec):
```
would `ceiling(prec)` work here?

ishanaj replied:
Yes, `ceiling(prec)` does work here. Difference being, in the previous case `prec` comes out to be of class `int` but now it would be of class `sympy.core.numbers.Integers`.
Should I commit this change?

gxyd replied:
I guess, that would work okay.

gxyd commented:
It has be great if you put the explanation you have given in the PR description into a commit message as well.

ishanaj committed:
incorporated ceiling() function

gxyd commented on sympy/polys/ring_series.py:
```diff
@@ -1140,6 +1140,8 @@ def rs_exp(p, x, prec):
     n = 1
     k = 1
     c = []
+    if not isinstance(prec, int):
```
no use of this check here.

gxyd commented on sympy/polys/ring_series.py:
```diff
@@ -1140,6 +1140,8 @@ def rs_exp(p, x, prec):
     c = []
+    if not isinstance(prec, int):
+        prec = ceiling(prec)
     for k in range(prec):
```
Can we simply do `for k in range(ceiling(k)):`?

gxyd replied:
I mean `ceiling(prec)` ofcourse, not `ceiling(k)`.

ishanaj replied:
Yes. I just checked for the same. It works.

ishanaj committed:
used ceiling() within for loop's range declaration

shivamvats commented:
All the basic algorithms in ringseries are meant to work with Taylor series (+ve integer exponents); see [this](https://github.com/sympy/sympy/blob/master/sympy/polys/ring_series.py#L22). Hence, they are not expected to handle a fractional `prec`. That happens in `rs_puiseux`.  It seems to me that either `rs_puiseux` is not working in Python3 or it is not being called (probably because of a difference in the  behaviour of division).

ishanaj commented:
@shivamvats On changing the last argument of the function `_rs_series()` in [this](https://github.com/sympy/sympy/blob/master/sympy/polys/ring_series.py#L1913) line to just `prec` the issue disappears for both python 2 and 3.
It seems to me that the argument `prec - sum_pows +min_pows[i])` was just to reduce further calculations(thus reducing the load on `rs_trunc()`-->[here](https://github.com/sympy/sympy/blob/master/sympy/polys/ring_series.py#L1990)).
But in this particular issue the expansion `exp(a + b**QQ(6,7))` comes out to be `exp(b**(6,7))`, while it should be `a*exp(b**(6,7)) + exp(b**(6,7))`, therefore reducing a term from the series.

ishanaj committed:
last argument of _rs_series() changed to prec

https://github.com/sympy/sympy/blob/master/sympy/polys/ring_series.py#L1913

On changing the last argument of the function _rs_series() in the above line to just prec the issue disappears for both python 2 and 3.
The argument (prec - sum_pows +min_pows[i]) was just to reduce further calculations(thus reducing the load on rs_trunc().
But in this particular issue, the expansion of exp(a + b**QQ(6,7)) comes out to be exp(b**(6,7)), while it should be a*exp(b**(6,7)) + exp(b**(6,7)), therefore reducing a term from the series.

shivamvats commented:
@ishanaj Does that happen in Python2 also?
I am getting the correct result:
```
In [6]: rs_series(exp(a+b**QQ(6,7)), a, 2)
Out[6]: (exp(b**(6/7)))*a + (exp(b**(6/7)))
```

ishanaj commented:
@shivamvats Individual evaluation comes out to be correct i.e`rs_series(sin(a**QQ(3,7)))`  and `rs_series(exp(a + b**QQ(6,7)), a,2)` gives the correct result in both python versions.

**But**, while evaluating `rs_series(sin(a**QQ(3,7))*exp(a + b**QQ(6,7)), a,2)` (which is the issue) [expr.is_Mul](https://github.com/sympy/sympy/blob/master/sympy/polys/ring_series.py#L1901) splits the whole expression in two arguments of multiplication, evaluates them separately and later multiplies the two results [here](https://github.com/sympy/sympy/blob/master/sympy/polys/ring_series.py#L1918) to give the resultant series expansion.
During this individual evaluation the term goes missing for `exp(a+b**QQ(6,7)), a, 2)` which I wanted to state above.

ylemkimon commented:
Please rebase on sympy/master. Thank you!

Abdullahjavednesar commented:
@shivamvats can you please have a look.

czgdp1807 commented:
@smichr Your thoughts on this PR? Can this be merged?

czgdp1807 commented:
ping @gxyd @Yathartha22 Is this PR ready to go. LGTM.

czgdp1807 commented:
ping @gxyd @ylemkimon @Yathartha22

Jatinbhardwaj-093 commented:
This can be closed as it got fixed in #27702
